cilium/CFP-30984-dns-proxy-ha.md

```diff
@@ -0,0 +1,151 @@
+
+### RPC Methods
+
+Method : UpdateMappings
+
+Request : 
+```
+message FQDNMapping {
+    string FQDN = 1;
+    repeated bytes IPS = 2;
+    int32 TTL = 3;
+    bytes client_ip = 4;
+}
+```
+Response : 
+```
+message Result {
+    bool success = 1;
+}
+```
+
+Method : UpdatesDNSRules
+
+_rpc UpdatesDNSRules(stream DNSPolicyRules) returns (Result){}_
+
+Request : 
+```
+message FQDNSelector {
+  string match_name = 1;
+  string match_pattern = 2;
+}
+
+message DNSPolicyRule {
+  uint32 port = 1;
+  string selector_string = 2;
+  repeated FQDNSelector port_rules= 3;
+  repeated string match_labels = 4;
+}
+
+message DNSPolicyRules {
+  uint64 endpoint_id = 1;
+  repeated DNSPolicyRule rules = 2;
+}
+```
+Response : 
+```
+message Result {
+    bool success = 1;
+}
+```
+
+Method : GetEndpointMetadataByIP
+
+Request : 
+```
+message FQDNSelector {
+  string ip = 1;
+}
+```
+Response : 
+```
+message Result {
+    uint32 endpoint_id = 1;
+    uint32 identity = 2;
+}
+```
+
+### Tracking policy updates to SDP instances
+
+Since SDP can be deployed out of band and users can choose to completely disable built-in proxy to run multiple instances of SDP, agent should be prepared to handle multiple instances. In order to ensure all instances  have upto date policy revisions, agent will maintain a mapping of ACKed policy revision numbers against stream ID.
+
+
+## Impacts / Key Questions
+
+
+### Discovering endpoint metadata from SDP
+
+In order to enforce L7 DNS policy, SDP needs to translate IP address to endpoint ID and identity ID. The simplest option is to reuse the same gRPC stream and implement another method to get the mappings from cilium agent. Since the actual source of truth are the bpf maps. Should the SDP be allowed to directly read from bpf maps ?
+
+
+### Option 1: gRPC method
+
+#### Pros
+
+* Avoid interacting with low-level details. Simpler to maintain and stay upto date with any datapath changes.
+* All data can be cached in-memory, in an event of agent being unavailable, SDP can lookup from in-memory cache
+
+#### Cons
+
+* Reliance on gRPC call in the hot-path
+* In an event where SDP restarts when agent's gRPC service is unavailable, all cached state is lost and SDP cannot translate IP to endpoint ID or identity.
+
+### Option 2: Read from bpf maps
+
+#### Pros
+
+* Reduced resource utilization in agent since agent doesn't need to process rpc calls in hotpath
```

Comment:
Is this true? A key feature of the DNS proxy is learning DNS -> IP mappings and informing the policy engine about them, so that the policy engine inside the cilium-agent can calculate policy impacts, allocate identities for those IPs, and populate BPF policymaps based on the new identities. If we do not ensure policy is plumbed before releasing DNS responses to the clients, then there is a high chance to automatically impose ~1s latency on subsequent TCP connections, because the first SYN of the connection will be dropped due to policy and the networking stack will delay a second before retrying.

(We could mitigate this in a number of ways, but I think that if we explore the use of inter-process RPC in hotpath holistically then it may not be the suboptimal path)

Reply:
iiuc, update dns->ip mappings not changing and  dnsproxy will be updating agent about new mappings via grpc and agent does all policy enforcement and only after getting ack from agent dns response will be returned to client. Reading from bpf maps/statefile option is for getting endpoint info and  ip->identity mapping to enforce dns policies. Mainly for dnsproxy to get reconcile on startup without relying on agent. @hemanthmalla can correct if i'm wrong.

Reply:
iiuc, there is no change to policy computation by agent. dns->ip mappings will be updated to agent via grpc api and agent does policy enforcement and only after getting ack from agent, dns response is returned back to client. Reading bpf maps is mainly for dnsproxy to get reconcile with ip->identity mappings and dnsproxy to bootup without relying on agent. @hemanthmalla correct me if i'm wrong

Reply:
I'm probably not reading the "agent doesn't need to process rpc calls in the hotpath" with the same set of assumptions then. @tamilmani1989 your description makes sense to me for the typical path when newly learned IP mappings are discovered, and for that specific scenario it sounds like the hotpath to me.

Reply:
@joestringer I agree with you WRT the RPC method to update DNS<>IP mappings. gRPC overhead for this method will likely be low relative to the computation needed to process the request.

But I intended this section to discuss how SDP would discover metadata to enforce L7 DNS policy and requests in this context are rpc calls to fetch identity and endpoint mappings. If we fetch those directly from bpf I was thinking we can skip a couple of rpc calls.

Reply:
Ah right. Yeah in terms of network handling latency the order of delay is likely to be local cache < syscall < rpc.